ptf/tests/linerate/meter_tests.py

```diff
@@ -0,0 +1,654 @@
+                ue_addr=UE1_ADDR, teid=UE1_UL_TEID,
+                pg_id=pg_id_ue1, l1_bps=stream_bps_ue1, l2_size=1400
+            ),
+            self.create_gtp_stream(
+                ue_addr=UE2_ADDR, teid=UE2_UL_TEID,
+                pg_id=pg_id_ue2, l1_bps=stream_bps_ue2, l2_size=1400
+            )
+        ]
+
+        self.trex_client.add_streams(streams, ports=TREX_OUT_PORT)
+        print(f"Starting traffic, duration: {TRAFFIC_DURATION_SECONDS} sec")
+        self.trex_client.start(TREX_OUT_PORT, duration=TRAFFIC_DURATION_SECONDS)
+        live_stats = monitor_port_stats(self.trex_client)
+        self.trex_client.wait_on_traffic(ports=TREX_OUT_PORT, rx_delay_ms=100)
+        avg_live_stats = self.average_monitored_port_stats(live_stats)
+
+        # Get and print TREX stats
+        trex_stats = self.trex_client.get_stats()
+        lat_stats_ue1 = get_latency_stats(pg_id_ue1, trex_stats)
+        lat_stats_ue2 = get_latency_stats(pg_id_ue2, trex_stats)
+        flow_stats_ue1 = get_flow_stats(pg_id_ue1, trex_stats)
+        flow_stats_ue2 = get_flow_stats(pg_id_ue2, trex_stats)
+
+        rx_bps_ue1 = (flow_stats_ue1.rx_bytes * 8) / TRAFFIC_DURATION_SECONDS
+        rx_bps_ue2 = (flow_stats_ue2.rx_bytes * 8) / TRAFFIC_DURATION_SECONDS
+
+        print("============= STATS UE 1 =============")
+        print(f"   RX Mbps: {rx_bps_ue1/M}")
+        print(get_readable_flow_stats(flow_stats_ue1))
+        print(get_readable_latency_stats(lat_stats_ue1))
+        print(SEPARATOR)
+        print("============= STATS UE 2 =============")
+        print(f"   RX Mbps: {rx_bps_ue2/M}")
+        print(get_readable_flow_stats(flow_stats_ue2))
+        print(get_readable_latency_stats(lat_stats_ue2))
+        print(SEPARATOR)
+
+        # No dropped packets for conforming traffic
+        self.assertEqual(
+            lat_stats_ue1.dropped,
+            0,
+            "Conforming traffic shouldn't get packet drops"
+        )
+        # Dropped packets for flow over the application meter rate
+        self.assertGreater(
+            lat_stats_ue2.dropped,
+            0,
+            "Misbehaving traffic should get packet drops"
+        )
+        self.assertAlmostEqual(
+            rx_bps_ue1 / session_rate,
+            1,
+            delta=0.05,
+            msg="UE 1 traffic should be almost equal to the session rate"
+        )
+        self.assertAlmostEqual(
+            rx_bps_ue2 / session_rate,
+            1,
+            delta=0.05,
+            msg="UE 2 traffic should be almost equal to the session rate"
+        )
+        self.assertAlmostEqual(
+            # 2 Flows, we expect twice the expected rate
+            avg_live_stats["avg_rx"][TREX_IN_PORT] / (2 * session_rate),
+            1,
+            delta=0.1,
+            msg="Traffic should be almost equal to twice the session rate"
+        )
+
+
+@group("trex-sw-mode")
+@group("upf")
+@group("meter")
+class ColorAwareSliceMeter(UpfPolicingTest):
+    """
```

Comment:
This description is listing the test steps, but the high-level intent is not clear. How about we simply say "Verifies that traffic above the session rate does not consume slice bandwidth.

Reply:
This makes me think that color-ware metering is an implementation detail, we are effectively testing some sort of fairness guarantee. Perhaps we could rename the test class into `UpfSliceFairPolicingTest`?

Reply:
I had some doubts also while writing the tests. Can we say anything about fairness? I believe, that since we are using a meter, we can't really assess fairness. Thus, I'm not sure we should verify the single flow rate.
I'm saying this because I believe that fairness would depend on the order in which packets are sent by TREX.

Reply:
I think it's ok to say that "it is NOT fair for RED packets (to be dropped) to consume slice bandwidth". Fariness is a generic/underspecified term so I'm ok using it. But if you prefer color-aware I'm fine too If you can use a consistent naming, e.g., `UpfSliceColorAwarePolicingTest`. Please note first comment about replacing the pydoc.